### PR TITLE
Page Editor pane to configure mod variables definition

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Support for editing mod variables definition in the Page Editor (behind feature flag): https://www.notion.so/pixiebrix/Configure-mod-variable-definitions-in-the-Page-Editor-065be18855e44797a38cc3893fe88cc9?pvs=4
- Feature flag: `page-editor-mod-variables-definition`
- Bug fixes:
  - 🐛  `isOutputKey` was missing start/end of text anchors
  - 🐛 Adapter method to select mod component left off mod variables definition
- Performance fixes
  - Memoize `selectGetDraftFormStatesPromiseForModId` 

## Demo

- https://www.loom.com/share/d15ca5b676ea437e93e4daf3516fb1bd?sid=349a4315-9a94-4e4a-b9f8-9c941bff6b90

## Remaining Work

- [x] Fix type errors
- [x] Place behind feature flag
- [x] Fix styling
- [x] Add tests
- [x] Add E2E tests

## Future Work

- Improve UX/UI layout
- Use mod variable declaration in variable popover. (Could be superset of what's found via analysis)
- Warn if a mod variable that's used is not declared (to encourage declaring mod variables)
- Warn if a mod variable that's declared is not used
- Find uses of a mod variable

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

File: src/components/tabLayout/EditorTabLayout.tsx [hunk 1]
```diff
@@ -62,18 +67,20 @@ const EditorTabLayout: React.FC<{
...
-                {badgeCount && badgeVariant && (
-                  <Badge className={styles.badge} variant={badgeVariant}>
-                    {badgeCount}
-                  </Badge>
-                )}
-              </Nav.Link>
-            </Nav.Item>
-          ))}
+          {tabs
+            .filter((x) => x.flag == null || flagOn(x.flag))
```
Comment (twschiller) [inline src/components/tabLayout/EditorTabLayout.tsx:71]: Add feature flag check

File: src/components/form/Form.tsx [hunk 2]
```diff
@@ -36,9 +37,9 @@ export type OnSubmit<TValues = FormikValues> = (
   formikHelpers: FormikHelpers<TValues>,
 ) => void | Promise<unknown>;
 
-export type RenderBody = (state: {
+export type RenderBody<TValues = FormikValues> = (state: {
```
Comment (twschiller) [inline src/components/form/Form.tsx:40]: Make type safe

File: src/pageEditor/starterBricks/adapter.ts [hunk 3]
```diff
@@ -148,15 +146,14 @@ export function formStateToDraftModComponent(
...
 
-export function selectGetDraftFormStatesPromiseForModId(state: RootState) {
-  return async (modId: RegistryId) => {
-    const getDraftModComponentsForMod =
-      selectGetDraftModComponentsForMod(state);
-
-    return Promise.all(
-      getDraftModComponentsForMod(modId).map(async (x) =>
-        isModComponentBase(x) ? modComponentToFormState(x) : x,
+export const selectGetDraftFormStatesPromiseForModId = createSelector(
```
Comment (twschiller) [inline src/pageEditor/starterBricks/adapter.ts:149]: Flagged by useSelector development checks

File: src/analysis/analysisVisitors/pageStateAnalysis/modVariableSchemasVisitor.ts [hunk 4]
```diff
@@ -28,39 +28,39 @@ export type ModVariableSchemaResult = {
   /**
    * Statically known mod variable schemas.
```
Comment (grahamlangford) [inline src/analysis/analysisVisitors/pageStateAnalysis/modVariableSchemasVisitor.ts:29]:
```suggestion
   * Statically known mod variable schema properties.
```

File: src/pageEditor/starterBricks/adapter.ts (hunk 3 above)
Comment (grahamlangford) [reply src/pageEditor/starterBricks/adapter.ts:149]: Gotta be my favorite new feature

File: src/pageEditor/tabs/modVariablesDefinition/ModVariablesDefinitionEditor.test.tsx [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        );
+      },
+    });
+
+    await waitFor(() => {
+      expect(
+        screen.getByRole("cell", { name: "declaredAny" }),
+      ).toBeInTheDocument();
+    });
+  });
```
Comment (grahamlangford) [inline src/pageEditor/tabs/modVariablesDefinition/ModVariablesDefinitionEditor.test.tsx:109]: NIT: Prefer `findBy` > `waitFor` when awaiting an element:

```suggestion
    await expect(
        screen.findByRole("cell", { name: "declaredAny" }),
      ).resolves.toBeInTheDocument();
  });
```

File: src/pageEditor/tabs/modVariablesDefinition/ModVariablesDefinitionEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+      />
+    </tbody>
+  </Table>
+);
+
+const ModVariablesDefinitionEditor: React.FC = () => {
+  const dispatch = useDispatch();
+  const modId = useSelector(selectActiveModId);
+
+  assertNotNullish(modId, "No active mod id");
```
Comment (grahamlangford) [inline src/pageEditor/tabs/modVariablesDefinition/ModVariablesDefinitionEditor.tsx:259]: We do this in enough places that I wonder if we should create a `selectActiveModIdOrThrow` selector?

File: src/pageEditor/tabs/modVariablesDefinition/ModVariablesDefinitionEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+        }),
+      description: Yup.string(),
+      type: Yup.string().required("Required"),
+      isAsync: Yup.boolean().required("Required"),
+      syncPolicy: Yup.string().required("Required"),
+    }),
+  ),
+});
+
+const VariableTable: React.FC<{
```
Comment (grahamlangford) [inline src/pageEditor/tabs/modVariablesDefinition/ModVariablesDefinitionEditor.tsx:72]: NIT: extract component to it's own file

File: src/pageEditor/tabs/modVariablesDefinition/modVariablesDefinitionEditorHelpers.ts [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+      data: dataDefinition,
+      error: {
+        type: "object",
+      },
+    },
+    required: ["isLoading", "isFetching", "isSuccess", "isError"],
+  };
+}
+
+export function mapDefinitionToFormValues({
```
Comment (grahamlangford) [inline src/pageEditor/tabs/modVariablesDefinition/modVariablesDefinitionEditorHelpers.ts:96]: NIT: Can we break this up into smaller helpers? The amount of nesting makes it difficult to quickly read/reason about

File: src/store/modComponents/modInstanceUtils.ts [hunk 9]
```diff
@@ -43,7 +46,10 @@ import { type DeploymentMetadata } from "@/types/deploymentTypes";
  */
 type ModInstanceActivatedModComponent = SetRequired<
   ActivatedModComponent,
-  "definitions" | "integrationDependencies" | "permissions"
+  | "definitions"
+  | "variablesDefinition"
+  | "integrationDependencies"
+  | "permissions"
```
Comment (grahamlangford) [inline src/store/modComponents/modInstanceUtils.ts:52]: NIT: alphabetical
```suggestion
  | "definitions"
  | "integrationDependencies"
  | "permissions"
  | "variablesDefinition"
```

File: end-to-end-tests/tests/pageEditor/modVariablesDefinition.spec.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { expect, test } from "../../fixtures/testBase";
+// @ts-expect-error -- https://youtrack.jetbrains.com/issue/AQUA-711/Provide-a-run-configuration-for-Playwright-tests-in-specs-with-fixture-imports-only
+import { test as base } from "@playwright/test";
+
+test("add/save mod variable definition", async ({
```
Comment (grahamlangford) [inline end-to-end-tests/tests/pageEditor/modVariablesDefinition.spec.ts:22]: Test the reverse? Define the mod variable in the mod variable definitions form then use in a brick?

Comment (twschiller) [reply end-to-end-tests/tests/pageEditor/modVariablesDefinition.spec.ts:22]: > We should also test the reverse: define the mod variable in the mod variable definitions form then use in a brick

There's some Jest coverage for the combinations: https://github.com/pixiebrix/pixiebrix-extension/blob/eb66de4ba45738778f651907f4bd1b4a97b0c183/src/pageEditor/tabs/modVariablesDefinition/ModVariablesDefinitionEditor.test.tsx#L38-L38

And mod variable runtime: https://github.com/pixiebrix/pixiebrix-extension/blob/eb66de4ba45738778f651907f4bd1b4a97b0c183/end-to-end-tests/tests/runtime/modVariables/variableSync.spec.ts#L25-L25

**Review by grahamlangford — APPROVED** [review]
See e2e test comment

File: end-to-end-tests/tests/pageEditor/modVariablesDefinition.spec.ts (hunk 10 above)
Comment (grahamlangford) [reply end-to-end-tests/tests/pageEditor/modVariablesDefinition.spec.ts:22]: That's fair

**Pull request merged**